### PR TITLE
Feat cross scope resources

### LINKED ISSUE #195: Refactor: Cross-Scope Resource Deployments
## Description
Implement the concept of "Resource Scopes" to the Azure Provider, to enable cross-scope deployments. Azure outlines the need for this [here](https://docs.microsoft.com/en-us/azure/azure-resource-manager/templates/cross-scope-deployment?tabs=azure-powershell).

### What Are Resource Scopes?

Scopes are best understood by reviewing an Azure ResourceId: 

#### Azure ResourceId Example - KeyVault Secret
`"/subscriptions/0000000-0000-0000-0000-00000000000/resourceGroups/example-integration-rossmurr4y-seg-baseline/providers/Microsoft.KeyVault/vaults/vault-1234567/secrets/mgmt-baseline-secret-ConnectionKey"`

Each section in the above resourceId is a different scope that this particular resource exists in. It has:
* a subscription of `0000000-0000-0000-0000-00000000000`
* a resourceGroup of `example-integration-rossmurr4y-seg-baseline`
* a provider of `Microsoft.KeyVault`
* a root parent resource
  * of the type `vaults`
  * with a name of `vault-1234567`
* finally it defines itself
  * with a type itself of `secrets`
  * and a name of `mgmt-baseline-secret-ConnectionKey`
  
It is important to understand that each segment of the resourceId is itself a new scope for that resource. However, for the remainder of this explanation we are only going to talk about the `subscription` and `resourceGroup` scopes.

### Scopes at Deployment Time
At the time of deployment, an ARM Template can be either deployed to a "ResourceGroup" scope (indicating that all resources contained in the template are deploying to a specified resourceGroup) or to a "Deployment" scope, indicating that all resources within the arm template exist above the scope of "ResourceGroup" - resources such as an "Account" only exist at this level - there is no "ResourceGroup" deployment for an "Account". This functionality already exists within the Hamlet `manageDeployment.sh` script in the Engine.

To look at it another way, a deployment of an ARM template at the `resourceGroup` scope will typically look like this:

```sh
az login

<login to an Azure Account and Subscription> 

az deployment group create --resource-group "my-example-resource-group" --template-file ./template.json
```

This will set the scope of all resources contained within the associated template to the `subscription` scope of the logged in account, and the `resourceGroup` scope of "my-example-resource-group".

### Scopes within an ARM Template
Because of the assignment of these two scopes at deployment time, all resources within an ARM template must be capable of their assignment. However, there are at least two common ways that this may not be a possibility. 

#### The Problems
1. a Resource within an ARM template simply does not exist at the Resource Group scope, where the rest of the resources in the template do.

2. A Resource has a parent that already exists within another scope.

#### Azure's Provided Solution

Azure provide a resource specifically for this purpose called  the "Deployment Resource". This resource type effectively defines additional deployments with alternate scopes within the one ARM Template. The "Deployment Resource" has a typical resource structure like any other, however through the use of a unique `template` property, an entire ARM Template can be "nested" within the resource. 

<details>
<summary>KeyVault Secret Resource (parent in the same resourceGroup)</summary>
<br>

```json
    {
      "name": "vault-1234567/exampleSecret",
      "type": "Microsoft.KeyVault/vaults/secrets",
      "apiVersion": "2018-02-14",
      "properties": {
         ...
      },
      "dependsOn": [
        ...
      ]
    }
```

</details>

<details>
<summary>KeyVault Secret Resource (nested inside a Deployment Resource)</summary>
<br>

```json
{
      "name": "exampleDeploymentResource",
      "type": "Microsoft.Resources/deployments",
      "apiVersion": "2019-10-01",
      "properties": {
        "template": {
          "$schema": "https://schema.management.azure.com/schemas/2019-04-01/deploymentTemplate.json#",
          "contentVersion": "1.0.0.0",
          "parameters": {},
          "resources": [
            {
              "name": "exampleSecret",
              "type": "Microsoft.KeyVault/vaults/secrets",
              "apiVersion": "2018-02-14",
              "properties": {
                ...
              }
            }
          ],
          "outputs": {
            "secretXfake": {
              "type": "string",
              "value": "[resourceId('Microsoft.KeyVault/vaults/secrets', 'rosstest')]"
            },
            "secretXfakeXname": {
              "type": "string",
              "value": "rosstest"
            }
          }
        }
      }
    }
```

</details>

### PR DESCRIPTION
**Pull request by RossMurr4y**
## Description
Implements Resource Scopes, explained in-depth in #195.

* Scopes are defined on a ResourceProfile, which outlines the lowest scope to which the resource will deploy to.
* Scopes are evaluated during the `@armResource` macro, determining the appropriate scope for the resource by:
   * Checking for parent scopes and using those
   * ensuring the scope does not go below the one defined by the ResourceProfile
* The `@armResource` macro will then either nest the resource to deploy it at the applicable scope, or decide no action is required.
* Should the scope dictate that a nested resource is required, the Outputs of the nested template are referenced by the outputs of the top-level template without additional configuration

### Additional Changes Included
* Moved the "Deployment" resource to a location that it is always loaded, as it is required for nesting by every Component.
* Refactored "pseudo" resource profiles - instead of shoehorning the idea of a "pseudo" resource into a unique resource type, "pseudo" is instead defined as a scope that sits above all others. A scope of "pseudo" allows such resources to be defined and used within Hamlet components just like any other, and yet only generate what is required.
* Refactored outputs so they can be constructed for any scope.

## Motivation and Context

* Closes #195
* Contributes to #196 

## How Has This Been Tested?

Local template generation

## Types of changes

- [ ] Bug fix (non-breaking change which fixes an issue)
- [x] Refactor (non-breaking change which improves the structure or operation of the implementation)
- [ ] New feature (non-breaking change which adds functionality)
- [ ] Breaking change (fix or feature that would cause existing functionality to change)

## Followup Actions

* Requires https://github.com/hamlet-io/engine-plugin-azure/pull/198

## Checklist:

- [ ] My change requires a change to the [documentation](https://github.com/hamlet-io/docs).
- [ ] I have updated the documentation accordingly.
- [ ] I have added tests to cover my changes.
- [ ] All new and existing tests passed.
- [ ] None of the above.

### CODE REVIEW

File: azure/deploymentframeworks/arm/output.ftl [hunk 1]
```diff
@@ -37,19 +37,63 @@
...
+    [#list mappings as attributeType,attributes]
+        [#local dataType = getOutputMappingDataType(attributeType)]
+        [#list attributes as attributeName,attributeValue]
+
+            [#switch scope]
+
+                [#case "subscription"]
+                [#case "resourceGroup"]
+                    [#if attributeValue == "id"]
+                        [#local typeFull = getAzureResourceProfile(AZURE_DEPLOYMENT_RESOURCE_TYPE).type]
```
Comment (roleyfoley) [inline azure/deploymentframeworks/arm/output.ftl:61]: Will this handle a missing resource profile?

File: azure/deploymentframeworks/arm/output.ftl [hunk 2]
```diff
@@ -37,19 +37,63 @@
...
+                        [#local propertySections = attributeValue?split(".")]
+                        [#local outputName = id]
+                        [#local value = getReference(id, name, typeFull, "", "", "", true, "outputs." + outputName + ".value")]
+                    [#else]
+                        [#local typeFull = getAzureResourceProfile(AZURE_DEPLOYMENT_RESOURCE_TYPE).type]
```
Comment (roleyfoley) [inline azure/deploymentframeworks/arm/output.ftl:66]: Will this handle a missing resource profile?

File: azure/deploymentframeworks/arm/output.ftl [hunk 3]
```diff
@@ -128,6 +172,117 @@
...
+            [#break]
+    [/#switch]
+[/#function]
+
+[#-- Deconstruct a resourceId value into discovered scopes --]
+[#function getResourceScopeFromResourcePath id]
+    [#local segments = getAzureResourcePropertySegments(id)]
+    [#local subscriptionIndex = segments?seq_index_of("subscriptions")!""]
+    [#local resourceGroupIndex = segments?seq_index_of("resourceGroups")!""]
+    [#local providerIndex = segments?seq_index_of("providers")!""]
```
Comment (roleyfoley) [inline azure/deploymentframeworks/arm/output.ftl:192]: Does it matter that this will default to a different type?

**Review by roleyfoley — CHANGES_REQUESTED** [review]
A couple of minor syntax bits but looks nice and simple to hanlde the complex case of these different deployment requirements

File: azure/deploymentframeworks/arm/output.ftl (hunk 1 above)
Comment (RossMurr4y) [reply azure/deploymentframeworks/arm/output.ftl]: Updated in 53c149e to fail elegantly

File: azure/deploymentframeworks/arm/output.ftl (hunk 2 above)
Comment (RossMurr4y) [reply azure/deploymentframeworks/arm/output.ftl]: Updated in 53c149e to fail elegantly

File: azure/deploymentframeworks/arm/output.ftl (hunk 3 above)
Comment (RossMurr4y) [reply azure/deploymentframeworks/arm/output.ftl:202]: The defaults are there so as to not give an unhelpful freemarker error message if its missing, I'm not really sure how to default to an integer nicely (i.e without setting it to some arbitrary number). I've added some validation in 9146a1f - the defaulting to an empty string will get us past the thrown freemarker error, and then the validation will allow the calls to `@fatal` to resolve and display to the end user if necessary.

File: azure/deploymentframeworks/arm/output.ftl [hunk 4]
```diff
@@ -37,19 +37,73 @@
...
+                    [/#if]
+                    [#break]
+
+                [#case "pseudo"]
+                    [#return getArmOutput(name, "string", "pseudo")]
+                    [#break]
+
+                [#default]
+                    [@fatal
+                        message=""
```
Comment (ml019) [inline azure/deploymentframeworks/arm/output.ftl:95]: Shouldn't the message have some content? I'd expect an explanation of why its a fatal

File: azure/deploymentframeworks/arm/output.ftl [hunk 5]
```diff
@@ -37,19 +37,73 @@
...
+    [#return {
+        name : {
+            "type" : type,
+            "value" : value
+        } +
+        attributeIfContent("condition", condition)
+    }]
+[/#function]
+
+[#function constructArmOutputsFromMappings id name scope mappings=[]]
```
Comment (ml019) [inline azure/deploymentframeworks/arm/output.ftl:50]: This function could do with a bit more commenting to explain the key differences between the output types

**Review by ml019 — CHANGES_REQUESTED** [review]

File: azure/deploymentframeworks/arm/output.ftl (hunk 5 above)
Comment (RossMurr4y) [reply azure/deploymentframeworks/arm/output.ftl:50]: Yep, have commented on this in another PR - #202

File: azure/deploymentframeworks/arm/output.ftl (hunk 4 above)
Comment (RossMurr4y) [reply azure/deploymentframeworks/arm/output.ftl]: Ahh yep, I intended to come back to this and must have seen something shiny over yonder

Comment (ml019) [reply azure/deploymentframeworks/arm/output.ftl]: he he

**Review by ml019 — APPROVED** [review]
All issues addressed

**Pull request merged**